### PR TITLE
[TODOs] add `EventApplicationOverserviced` event

### PR DESCRIPTION
**Pull request by bryanchriswhite**
## Summary

Add `EventApplicationOverserviced` when application stake is less than the settlement amount so that we can monitor occurrences of this scenario and compute the effective inflation.

## Issue

While the TODO asks to "unstake" the application, such an action isn't precisely defined. When using the CLI, "unstaking" has the effect of moving staked tokens to the application address and then removing the application from the state. In the current context, there is no stake to move, leaving "unstake" to mean to remove the application from the state store. 

Doing this automatically as the result of an action other than explicitly unstaking seems likely to me to introduce confusion. For example, imagine an application querying the network to check their staked balance. If we remove the application at the end of a session where the application was overserviced, such queries would return an app not found error. This also has the effect of wiping out information about the application's on-chain service configuration, which may be useful as an intent signal.

Once #612 is implemented, **unintentional** application overservicing will be mitigated. An attack vector remains which is not trivially addressed; "unstaking" overserviced application has a negligible impact on this while coming with the aforementioned trade-offs. 

- #584

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

**Documentation changes** (only if making doc changes)
- [ ] `make docusaurus_start`; only needed if you make doc changes

**Local Testing** (only if making code changes)
- [ ] **Unit Tests**: `make go_develop_and_test`
- [ ] **LocalNet E2E Tests**: `make test_e2e`
- See [quickstart guide](https://dev.poktroll.com/developer_guide/quickstart) for instructions

**PR Testing** (only if making code changes)
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.
    - **THIS IS VERY EXPENSIVE**, so only do it after all the reviews are complete.
    - Optionally run `make trigger_ci` if you want to re-trigger tests without any code changes
    - If tests fail, try re-running failed tests only using the GitHub UI as shown [here](https://github.com/pokt-network/poktroll/assets/1892194/607984e9-0615-4569-9452-4c730190c1d2)

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [x] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

## Summary by CodeRabbit

- **New Features**
  - Introduced `EventApplicationOverserviced` message to track application over-service events, including expected and effective burns.

- **Refactor**
  - Updated package imports and type declarations to align with the new naming conventions for better code readability and organization.
  - Modified error return statements for improved error handling consistency.

- **Tests**
  - Added event attribute validation and coin unmarshalling to the session accounting tests for enhanced test coverage.

### CODE REVIEW

File: x/tokenomics/keeper/settle_session_accounting.go [hunk 1]
```diff
@@ -150,24 +150,37 @@ func (k Keeper) SettleSessionAccounting(
...
 		// Do we just let it go into debt? Do we penalize the application? Do we unstake it? Etc...
-		settlementAmt = sdk.NewCoin("upokt", math.Int(application.Stake.Amount))
-		settlementAmtuPOKT = sdk.NewCoins(settlementAmt)
-		// TODO_BLOCKER(@Olshansk): The application should be immediately unstaked at this point in time
+		settlementAmt = cosmostypes.NewCoin("upokt", math.Int(application.Stake.Amount))
+		settlementAmtuPOKT = cosmostypes.NewCoins(settlementAmt)
+
+		applicationOverservicedEvent := &tokenomicstypes.EventApplicationOverserviced{
+			ApplicationAddr: applicationAddress.String(),
+			ExpectedBurn:    &settlementAmt,
```
Comment (red-0ne) [inline x/tokenomics/keeper/settle_session_accounting.go:158]: `settlementAmt` has already be overridden by `application.Stake.Amount` at this point.

File: x/tokenomics/keeper/settle_session_accounting.go [hunk 2]
```diff
@@ -150,24 +150,37 @@ func (k Keeper) SettleSessionAccounting(
...
+			EffectiveBurn:   application.GetStake(),
+		}
+		if err = cosmostypes.UnwrapSDKContext(ctx).EventManager().EmitTypedEvent(applicationOverservicedEvent); err != nil {
```
Comment (red-0ne) [inline x/tokenomics/keeper/settle_session_accounting.go:161]:
```suggestion
		eventManager := cosmostypes.UnwrapSDKContext(ctx).EventManager()
		if err = eventManager.EmitTypedEvent(applicationOverservicedEvent); err != nil {
```

File: x/tokenomics/keeper/settle_session_accounting.go (hunk 1 above)
Comment (bryanchriswhite) [reply x/tokenomics/keeper/settle_session_accounting.go:158]: Can you say more, I'm not quite following. :thinking:

Comment (red-0ne) [reply x/tokenomics/keeper/settle_session_accounting.go:158]: `settlementAmt` is overridden at line `153` to `application.Sake.Amount` so it's no longer the expected amount to burn.

At this point in code, `settlementAmt` and `application.Stake` are the same. What we should do is to `settlementAmt` to a `expectedSettleemntAmt` before reassigning `settlementAmt` to `application.Stake.Amount` (`L153`) then use the new variable in `ExpectedBurn`.

**Review by red-0ne — APPROVED** [review]
LGTM!

File: x/tokenomics/keeper/settle_session_accounting.go [hunk 3]
```diff
@@ -29,40 +29,40 @@ func (k Keeper) SettleSessionAccounting(
 ) error {
 	logger := k.Logger().With("method", "SettleSessionAccounting")
 
-	settlementAmt := sdk.NewCoin("upokt", math.NewInt(0))
+	settlementCoin := cosmostypes.NewCoin("upokt", math.NewInt(0))
```
Comment (Olshansk) [inline x/tokenomics/keeper/settle_session_accounting.go:32]: Thoughts on pluralizing this to `Coins`?

File: x/tokenomics/keeper/settle_session_accounting.go [hunk 4]
```diff
@@ -114,60 +114,76 @@ func (k Keeper) SettleSessionAccounting(
...
 		))
 		// TODO_MAINNET(@Olshansk, @RawthiL): The application was over-serviced in the last session so it basically
 		// goes "into debt". Need to design a way to handle this when we implement
 		// probabilistic proofs and add all the parameter logic. Do we touch the application balance?
 		// Do we just let it go into debt? Do we penalize the application? Do we unstake it? Etc...
-		settlementAmt = sdk.NewCoin("upokt", math.Int(application.Stake.Amount))
-		settlementAmtuPOKT = sdk.NewCoins(settlementAmt)
-		// TODO_BLOCKER(@Olshansk): The application should be immediately unstaked at this point in time
+		expectedBurn := settlementCoin
+		settlementCoin = *application.GetStake()
```
Comment (Olshansk) [inline x/tokenomics/keeper/settle_session_accounting.go:154]:
```suggestion
		// Make the settlement amount the maximum stake that the application has remaining
		settlementCoin = *application.GetStake()
```

**Review by Olshansk — APPROVED** [review]

File: x/tokenomics/keeper/settle_session_accounting.go (hunk 3 above)
Comment (bryanchriswhite) [reply x/tokenomics/keeper/settle_session_accounting.go:32]: `settlementCoins` is used elsewhere and is a `[]cosmostypes.Coin` type.

**Pull request merged**